### ISSUE TITLE
Allow styling custom styles registered via register_block_style in theme.json

### ISSUE THREAD

**Issue by mrleemon** [WordPress/gutenberg#49550]
Since WP 6.2, one can style core block style variations via the "variations" key in theme.json that can be added under styles.blocks.blockname. Right now, this does not work with custom styles registered via register_block_style.

Can you add this feature, please?

**Comment by t-hamano** [comment]
@mrleemon 

Thanks for the report.

Currently, #49396 has been submitted to add custom CSS to elements and variations. I believe this PR will solve your problem.

**Comment by mrleemon** [comment]
Sorry, I didn't see that PR. Thanks!

**Comment by mrleemon** [comment]
Reopening because the linked PR is about adding support for the "css" property to elements and variations, and my issue is about adding simple styling support (color, margin, padding, typography, etc.) to variations.

**Comment by t-hamano** [comment]
Regarding adding styling support to block variations, I think this was implemented in #46343. The example code can be found in [this section](https://make.wordpress.org/core/2023/03/07/miscellaneous-editor-changes-in-wordpress-6-2/#edit-block-style-variations-from-theme-json), which is part of the WordPress 6.2 Field Guide.

**Comment by mrleemon** [comment]
> Regarding adding styling support to block variations, I think this was implemented in #46343. The example code can be found in [this section](https://make.wordpress.org/core/2023/03/07/miscellaneous-editor-changes-in-wordpress-6-2/#edit-block-style-variations-from-theme-json), which is part of the WordPress 6.2 Field Guide.

Yes, I know, but it only works for CORE styles (for example, the "rounded" style in the image block), not custom styles registered via `register_block_style`

**Comment by tellthemachines** [comment]
Thanks for bringing this up!

> it only works for CORE styles (for example, the "rounded" style in the image block), not custom styles registered via register_block_style

Correct, the initial implementation only included core style variations, with the standard design tools. Because existing custom style variations are styled via a classic stylesheet, which depending on specificity might not be overridable with the core design tools such as color, typography etc.., custom styles were excluded then. 

I'm thinking that the custom CSS field could be leveraged to solve that problem, and also that it should be possible to add new custom variations from theme.json. I opened an issue in #49602 suggesting some next steps.

**Comment by edmundcwm** [comment]
Isn't this already working from 6.6 onwards? Quote from this [article](https://developer.wordpress.org/block-editor/how-to-guides/themes/global-settings-and-styles/#styles): 
>WordPress 6.6 will let you style any registered block style variations in theme.json. Just like with Core variations, you can target your custom variations via the styles.blocks.blockName.variations property.

In the article example it uses `register_block_style()` to register a custom block style variation and style it via `theme.json` which I believe is what OP is asking.

**Comment by t-hamano** [comment]
I think you're right. Example:

theme.json

```json
{
	"version": 3,
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "840px",
			"wideSize": "1100px"
		}
	},
	"styles": {
		"blocks": {
			"core/button": {
				"variations": {
					"my-style": {
						"color": {
							"background": "#ff0000",
							"text": "#00ff00"
						},
						"typography": {
							"fontSize": "20px"
						}
					}
				}
			}
		}
	}
}
```

functions.php

```php
register_block_style(
	'core/button',
	[
		'name'  => 'my-style',
		'label' => 'My Style',
	]
);
```

Custom styles applied to the Button block

![Image](https://github.com/user-attachments/assets/7f5f7580-b0de-47e2-8ff6-e509e036211e)

Can change the style from the global styles

![Image](https://github.com/user-attachments/assets/74a03280-3cab-42f7-8339-184f7e66c928)

![Image](https://github.com/user-attachments/assets/7103d1b1-45f0-4a18-9a23-79557b0e7d4a)

@mrleemon I would like to close this issue again, but if this isn't what you were looking for, please feel free to comment.

**Comment by mrleemon** [comment]
This issue can be closed as the functionality is already working. But, it seems there's an error in the "Variations" section of the [docs](https://developer.wordpress.org/block-editor/how-to-guides/themes/global-settings-and-styles/#styles). The section incorrectly states that styles added via register_block_style are ignored, which is no longer accurate:

>It’s also worth highlighting that only variations defined in the block.json file of the block are considered “registered”: so far, the style variations added via register_block_style or in the client are ignored, see [this issue](https://github.com/WordPress/gutenberg/issues/49602) for more information.

**Comment by t-hamano** [comment]
> it seems there's an error in the "Variations" section of the [docs](https://developer.wordpress.org/block-editor/how-to-guides/themes/global-settings-and-styles/#styles). The section incorrectly states that styles added via register_block_style are ignored, which is no longer accurate:
> 
> >It’s also worth highlighting that only variations defined in the block.json file of the block are considered “registered”: so far, the style variations added via register_block_style or in the client are ignored, see [this issue](https://github.com/WordPress/gutenberg/issues/49602) for more information.

I'm not sure if the issue mentioned above has been completely resolved, but the variations registered via `register_block_style` do indeed seem to work as tested in [this comment](https://github.com/WordPress/gutenberg/issues/49550#issuecomment-2541552985).

@aaronrobertshaw @tellthemachines Do you know if it's possible to remove this warning from the doc entirely?

**Comment by mrleemon** [comment]
Yes, the issue linked in the quote is still unresolved, but the one regarding `register_block_style` has been fixed.

**Comment by aaronrobertshaw** [comment]
> Do you know if it's possible to remove this warning from the doc entirely?

@t-hamano updating the docs here has been on my radar but I had other priorities come up.

Removing the warning entirely would probably be insufficient. Instead, we'll need to fix the flagged issue and alter the rest of it to be more nuanced. 

> Note that variations are a “block concept”, they only exist bound to blocks. The theme.json specification respects that distinction by only allowing variations at the block-level but not at the top-level. It’s also worth highlighting that only variations defined in the block.json file of the block are considered “registered”: so far, the style variations added via register_block_style or in the client are ignored, see [this issue](https://github.com/WordPress/gutenberg/issues/49602) for more information.

Block style variations are a "block concept" and it should still be conveyed they are bound to blocks. To style a block style variation via theme.json (e.g. via a theme style variation or through a theme.json partial), the block style needs to have been registered. 

IIRC global styles changes to block style variation styles will require that the block style variation was registered via `register_block_style` server-side. If a block style is only registered via JS when the theme.json is sanitized (especially for users without unfiltered HTML caps), the JS-only registered block style data will be stripped out.

Lastly, due to the ability to apply block style variations in a nested manner, the output CSS in the docs is currently incorrect too.  Selectors for block style variations will now be given a unique class to facilitate the nesting e.g. `is-style-dark-123` as opposed to simply `is-style-dark`.

The TL;DR is, yes, the docs need updating. I'll try and put something together in the coming days that at least fixes the glaring issues currently in the docs.

**Comment by edmundcwm** [comment]
> IIRC global styles changes to block style variation styles will require that the block style variation was registered via register_block_style server-side. If a block style is only registered via JS when the theme.json is sanitized (especially for users without unfiltered HTML caps), the JS-only registered block style data will be stripped out

I guess this explains why when I tested out using `register_block_style` on the server side and `registerBlockStyle` on the client, only the former block style allows styling via `theme.json`.

Thanks for the clarification @aaronrobertshaw !

**Comment by aaronrobertshaw** [comment]
I've thrown up a quick PR to update the block style variation docs in https://github.com/WordPress/gutenberg/pull/68169. Feel free to leave any feedback you have over on that PR.
